### PR TITLE
FEATURE: Push Notifications for Chrome.

### PR DESCRIPTION
**Pull request by tgxworld**
bug_report-19

- [x] Handle config of push notifications properly
- [x] Fallback to Desktop notification when push notification is disabled.
- [ ] Handle condition when connection is not done over HTTPS

### CODE REVIEW

File: app/controllers/push_notifications_controller.rb [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+      render json: failed_json
+    end
+  end
+
+  def unsubscribe
+    PushNotifications.unsubscribe(current_user)
+    render json: success_json
+  end
+
+  def latest
```
Comment (tgxworld) [inline app/controllers/push_notifications_controller.rb]: Payload hasn't been released for push messaging so we have to do this. See https://code.google.com/p/chromium/issues/detail?id=486040

File: lib/push_notifications.rb [hunk 2]
```diff
@@ -0,0 +1,51 @@
+class PushNotifications
+  KEY_PREFIX = "push-notification-subscription:".freeze
+  GCM_ENDPOINT = 'https://android.googleapis.com/gcm/send'.freeze
+
+  def self.registration_id(user)
+    $redis.get(redis_key(user))
+  end
+
+  def self.subscribe(user, endpoint)
+    $redis.set(redis_key(user), extract_registration_id(endpoint))
```
Comment (tgxworld) [inline lib/push_notifications.rb]: TODO: This is sharing the registration ID across browsers.

File: config/locales/server.en.yml [hunk 3]
```diff
@@ -2080,6 +2082,9 @@ en:
 
         If the above link is not clickable, try copying and pasting it into the address bar of your web browser.
 
+    push:
+      body: You have %{count} new notifications.
```
Comment (gschlager) [inline config/locales/server.en.yml]: This needs pluralization (`one` and `other`).

Comment (tgxworld) [inline config/locales/server.en.yml]: hmm I could add that in but currently, we won't call the translation when we only have one notification.

Comment (gschlager) [inline config/locales/server.en.yml]: It's needed, even if it doesn't make sense for English. Other languages have other pluralization rules and Transifex doesn't detect it as pluralized unless there are `one` and `other` keys in English. Trust me. ;-)

Comment (tgxworld) [inline config/locales/server.en.yml]: Ah ic. :+1: thanks!!

File: lib/push_notifications.rb [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        body = URI.encode_www_form(
+          registration_id: registration_id,
+          collapse_key: Discourse.base_url
+        )
+
+        result = Excon.post(GCM_ENDPOINT, headers: headers, body: body)
+
+        if result.status == 200 && result.body.start_with?("Error")
+          Rails.logger.error("There was an error from Google Cloud Message: #{result.body}")
+        elsif result.status != 200
```
Comment (tgxworld) [inline lib/push_notifications.rb]: TODO: Check the response status and unsubscribe registration ids that are no longer in use.

**Pull request closed without merging**